src/Asm89/Stack/CorsService.php

```diff
@@ -37,140 +37,104 @@ private function normalizeOptions(array $options = array())
 
         // normalize array('*') to true
-        if (in_array('*', $options['allowedOrigins'])) {
+        if ($options['allowedOrigins'] === array('*')) {
             $options['allowedOrigins'] = true;
         }
-        if (in_array('*', $options['allowedHeaders'])) {
+        if ($options['allowedHeaders'] === array('*')) {
             $options['allowedHeaders'] = true;
         } else {
             $options['allowedHeaders'] = array_map('strtolower', $options['allowedHeaders']);
         }
-
-        if (in_array('*', $options['allowedMethods'])) {
+        if ($options['allowedMethods'] === array('*')) {
             $options['allowedMethods'] = true;
         } else {
             $options['allowedMethods'] = array_map('strtoupper', $options['allowedMethods']);
         }
+        if ($options['exposedHeaders'] === array('*')) {
+            $options['exposedHeaders'] = true;
+        } elseif (is_array($options['exposedHeaders'])) {
+            $options['exposedHeaders'] = array_map('strtolower', $options['exposedHeaders']);
+        }
 
         return $options;
     }
 
-    public function isActualRequestAllowed(Request $request)
-    {
-        return $this->checkOrigin($request);
-    }
-
-    public function isCorsRequest(Request $request)
-    {
-        return $request->headers->has('Origin') && !$this->isSameHost($request);
-    }
-
     public function isPreflightRequest(Request $request)
     {
-        return $this->isCorsRequest($request)
-            && $request->getMethod() === 'OPTIONS'
+        return $request->getMethod() === 'OPTIONS'
             && $request->headers->has('Access-Control-Request-Method');
     }
 
     public function addActualRequestHeaders(Response $response, Request $request)
     {
-        if (!$this->checkOrigin($request)) {
-            return $response;
-        }
-
-        $response->headers->set('Access-Control-Allow-Origin', $request->headers->get('Origin'));
-
-        if (!$response->headers->has('Vary')) {
-            $response->headers->set('Vary', 'Origin');
-        } else {
-            $response->headers->set('Vary', $response->headers->get('Vary') . ', Origin');
-        }
+        $this->addOriginHeader($response, $request);
+        $this->addCredentialsHeader($response);
 
         if ($this->options['supportsCredentials']) {
             $response->headers->set('Access-Control-Allow-Credentials', 'true');
         }
 
         if ($this->options['exposedHeaders']) {
-            $response->headers->set('Access-Control-Expose-Headers', implode(', ', $this->options['exposedHeaders']));
+            if ($this->options['supportsCredentials'] === true && $this->options['exposedHeaders'] === true) {
+                // This is not allowed.
+                // @see https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Expose-Headers
+                throw new \LogicException("Cannot set supportsCredentials to true and exposedHeaders to true or ['*']");
```

Comment:
Should we throw this earlier? Because cors errors are swallowed by the browser.

Reply:
I moved this to `normalizeOptions()` or it could be in the constructor directly if you prefer.